### PR TITLE
Add spec wsn tests

### PR DESCRIPTION
**Pull request by chadpowers**
see [roadmap #434](https://github.com/kaazing/roadmap/issues/434)

### CODE REVIEW

**Comment by dpwspoon** [comment]
Minor feedback, good work @chadpowers.  Would one of @cmebarrow or @jitsni mind reviewing before merging?  Thanks

**Comment by dpwspoon** [comment]
Any more analysis you can do on the ignore to more precisely describe the issue would be useful.

File: transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/OpeningHandshakeIT.java [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+        })
+    public void shouldEstablishConnectionWithRandomCaseResponseHeaders() throws Exception {
+        k3po.finish();
+    }
+
+    @Test
+    @Specification({
+        "request.header.origin/handshake.request"
+        })
+    public void shouldEstablishConnectionWithRequestHeaderOrigin() throws Exception {
```
Comment (cmebarrow) [inline transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/OpeningHandshakeIT.java]: There should be a negative test for the case where the origin header value is not accepted by the server (which according to RFS 6455 section 10.2 should result in HTTP 403 Forbidden status code). This test needs to be added first to the wsn spec project.

File: transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/BaseFramingIT.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        }
+    };
+    
+    private final TestRule timeout = new DisableOnDebug(new Timeout(5, SECONDS));
+
+    @Rule
+    public final TestRule chain = outerRule(k3po).around(gateway).around(timeout);
+
+    @Test
+    @Ignore("Error: expected 'read [0x82 0x00]', actual 'Disconnected', Not echoing 0 length payload")
```
Comment (cmebarrow) [inline transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/BaseFramingIT.java]: If this is a bug please file one and reference it here

File: transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/BaseFramingIT.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    @Test
+    @Specification({
+        "echo.binary.payload.length.65536/handshake.request.and.frame"
+        })
+    public void shouldEchoBinaryFrameWithPayloadLength65536() throws Exception {
+        k3po.finish();
+    }
+
+    @Test
+    @Ignore("Error: expected 'read [0x82 0x00]', actual 'Disconnected', Not echoing 0 length payload")
```
Comment (cmebarrow) [inline transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/BaseFramingIT.java]: If this is a bug please file one and reference it here

**Comment by cmebarrow** [comment]
@chadpowers Please react to my review comments so we we move ahead

**Comment by chadpowers** [comment]
@cmebarrow @dpwspoon I am not sure exactly how you want me to address the @Ignore's. I do not know whether they are bugs with the script, the gateway, or my tests.

File: transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/ClosingHandshakeIT.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+    @Test
+    @Specification({
+        "client.send.close.frame.with.code.1000.and.reason/handshake.request.and.frame"})
+    public void shouldCompleteCloseHandshakeWhenClientSendCloseFrameWithCode1000AndReason() throws Exception {
+        k3po.finish();
+    }
+
+    @Test
+    @Ignore("Read code different than expected and did not fail")
```
Comment (jitsni) [inline transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/ClosingHandshakeIT.java]: What status code did gateway send ? It should be 1007 or 1002 or drop the connection.

File: transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/ClosingHandshakeIT.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    @Specification({
+        "client.send.close.frame.with.code.1000.and.invalid.utf8.reason/handshake.request.and.frame"})
+    public void shouldFailWebSocketConnectionWhenClientSendCloseFrameWithCode1000AndInvalidUTF8Reason() throws Exception {
+        k3po.finish();
+    }
+
+    @Test
+    @Ignore("Read code off by byte, not matching spec test")
```
Comment (jitsni) [inline transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/ClosingHandshakeIT.java]: I think spec test is looking for 1000, and gateway is just echoing back 1001. Is that correct ? Some impls
1. don't send any status code
2. send 1000
3. send 1001

Gateway can be modified to do 1 or 2, but also the spec test shouldn't be rigid.

File: transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/ControlIT.java [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+    @Test
+    @Specification({
+        "client.send.ping.payload.length.126/handshake.request.and.frame"
+        })
+    public void shouldFailWebSocketConnectionWhenClientSendPingFrameWithPayloadTooLong() throws Exception {
+        k3po.finish();
+    }
+
+    @Test
+    @Ignore("Timeout error")
```
Comment (jitsni) [inline transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/ControlIT.java]: Seems like a spec test issue. I wonder why it is expecting gateway to close the connection.

File: transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/ControlIT.java [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+        "client.send.pong.payload.length.0/handshake.request.and.frame"
+        })
+    public void shouldReceiveClientPongFrameWithEmptyPayload() throws Exception {
+        k3po.finish();
+    }
+
+    @Test
+    @Ignore("Timeout error")
+    @Specification({
+        "client.send.pong.payload.length.125/handshake.request.and.frame"
```
Comment (jitsni) [inline transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/ControlIT.java]: Spec issue. Wonder why it is expecting gateway to close the connection

File: transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/FragmentationIT.java [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+    @Test
+    @Specification({
+        "client.echo.text.payload.length.125.not.fragmented/handshake.request.and.frame"
+        })
+    public void shouldEchoClientSendTextFrameWithPayloadNotFragmented() throws Exception {
+        k3po.finish();
+    }
+
+    @Test
+    @Ignore("Gateway disconnected instead of echoing code, client did not read")
```
Comment (jitsni) [inline transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/FragmentationIT.java]: Gateway issue. We used to JIRA issue (just create an issue)

File: transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/FragmentationIT.java [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+        "client.echo.text.payload.length.0.fragmented/handshake.request.and.frames"
+        })
+    public void shouldEchoClientSendTextFrameWithEmptyPayloadFragmented() throws Exception {
+        k3po.finish();
+    }
+
+    @Test
+    @Ignore("Gateway writing wrong bytes back to client, read does not match")
+    @Specification({
+        "client.echo.text.payload.length.0.fragmented.with.injected.ping.pong/handshake.request.and.frames"
```
Comment (jitsni) [inline transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/FragmentationIT.java]: Same as above,

File: transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/FragmentationIT.java [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+        })
+    public void shouldEchoClientSendTextFrameWithEmptyPayloadFragmentedAndInjectedPingPong() throws Exception {
+        k3po.finish();
+    }
+
+    @Test
+    @Ignore("Incorrect data write by Gateway, not echoing")
+    @Specification({
```
Comment (jitsni) [inline transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/FragmentationIT.java]: Gateway just sends data in multiple fragments. Spec test provides one of the ways data is received but gateway(or any implementation) may not echo the data in the spec expected framing. Spec needs to provide different scripts(or some sort of websocket framing knowledge should be built in robot).

**Comment by jitsni** [comment]
For some of the tests, echo service doesn't do the job. You may need to wire WsnAcceptor and write custom logic (for e.g consolidate all the fragments of a message and echo one message)

File: transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/BaseFramingIT.java (hunk 2 above)
Comment (chadpowers) [inline transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/BaseFramingIT.java]: Filed [Bug](https://github.com/kaazing/gateway/issues/254)

File: transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/BaseFramingIT.java (hunk 3 above)
Comment (chadpowers) [inline transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/BaseFramingIT.java]: Filed [Bug](https://github.com/kaazing/gateway/issues/254)

File: transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/ClosingHandshakeIT.java (hunk 4 above)
Comment (chadpowers) [inline transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/ClosingHandshakeIT.java]: @jitsni the expected is this
```
read [0x88 0x02 0x03 0xea]

closed
```
and actual is
```
read [0x88 0x16 0x03 0xe8]
```

The gateway is not closing at the script says it should. Where should I file an issue?

File: transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/ClosingHandshakeIT.java (hunk 5 above)
Comment (chadpowers) [inline transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/ClosingHandshakeIT.java]: @jitsni 
Here expected is:
```
read [0x88 0x02 0x03 0xe8]

closed
```
but actual is 
```
read [0x88 0x02 0x03 0xe9]
```

Again, gateway is also not closing. Is this a script, gateway, or test login issue?

File: transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/ControlIT.java (hunk 6 above)
Comment (chadpowers) [inline transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/ControlIT.java]: Filed [spec bug](https://github.com/k3po/k3po/issues/208)

File: transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/ControlIT.java (hunk 7 above)
Comment (chadpowers) [inline transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/ControlIT.java]: Filed [spec bug](https://github.com/k3po/k3po/issues/208)

File: transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/FragmentationIT.java (hunk 10 above)
Comment (chadpowers) [inline transport/wsn/src/test/java/org/kaazing/gateway/transport/wsn/specification/ws/FragmentationIT.java]: Does this need connect/acceptor login similar to what I did for TCP in NIO?

See [here](https://github.com/kaazing/gateway/pull/235/files)

**Comment by chadpowers** [comment]
I added test logic to consolidate fragmentations. Now 15/20 tests in that IT pass.

**Comment by jitsni** [comment]
Created a new PR (https://github.com/kaazing/gateway/pull/270) using this one.

**Pull request closed without merging**